sync/notify/bugs.py

```diff
@@ -0,0 +1,228 @@
+            suffix = test_id.rsplit(".", 1)[0] + ".js"
+        elif test_id.endswith(".window.html"):
+            suffix = test_id.rsplit(".", 1)[0] + ".js"
+
+        path = prefix + suffix
+        data[path].append(test_id)
+    return data
+
+
+@mut('sync')
+def for_sync(sync, results):
+    # TODO: file bugs for things in the results that require a bug and don't have one
+    rv = {}
+
+    new_crashes = list(results.iter_filter(lambda _test, _subtest, result:
+                                           (result.has_crash("firefox") and
+                                            not result.has_link(status="CRASH"))))
+
+    new_failures = list(results.iter_filter(lambda _test, _subtest, result:
+                                            ((result.has_regression("firefox") or
+                                              result.has_new_non_passing("firefox") or
+                                              result.is_browser_only_failure("firefox")) and
+                                             not result.has_link())))
+
+    if not new_failures and not new_crashes:
+        return rv
+
+    git_work = sync.gecko_worktree.get()
+    path_prefix = env.config["gecko"]["path"]["wpt"]
+
+    seen = set()
+
+    for test_results, bug_data, link_status, require_opt_in in [
+            (new_crashes, bug_data_crash, "CRASH", False),
+            (new_failures, bug_data_failure, None, True)]:
+
+        # Tests excluding those for which we already generated a bug
+        test_results = [item for item in test_results
+                        if (item[0], item[1]) not in seen]
+
+        seen |= set((item[0], item[1]) for item in test_results)
+
+        test_ids = [test_id for test_id, _subtest, _result in test_results]
+        test_id_by_path = test_ids_to_paths(git_work, test_ids)
+        test_path_by_id = {}
+        for path, ids in iteritems(test_id_by_path):
+            for test_id in ids:
+                assert test_id not in test_path_by_id
+                test_path_by_id[test_id] = os.path.relpath(path, path_prefix)
+
+        components = components_for_wpt_paths(git_work,
+                                              set(itervalues(test_path_by_id)))
+
+        components_by_path = {}
+        for component, paths in iteritems(components):
+            for path in paths:
+                components_by_path[path] = component
+
+        test_results_by_component = defaultdict(list)
+
+        for test_id, subtest, test_result in test_results:
+            component = components_by_path[test_path_by_id[test_id]]
+            test_results_by_component[component].append((test_id, subtest, test_result))
+
+        for component, test_results in iteritems(test_results_by_component):
+            if component == "UNKNOWN":
+                # For things with no component don't file a bug
+                continue
+
+            if require_opt_in:
+                opt_in_components = [item.strip()
```

Comment:
bit nit picky, but wouldn't it be better to generate this once before the loop instead of during every iteration? This is assuming it doesn't change between iterations